beta/src/components/MDX/Sandpack/LoadingOverlay.tsx

```diff
@@ -0,0 +1,143 @@
+  OpenInCodeSandboxButton,
+  useSandpack,
+} from '@codesandbox/sandpack-react';
+import {useEffect} from 'react';
+
+const FADE_ANIMATION_DURATION = 200;
+
+export const LoadingOverlay = ({
+  clientId,
+  dependenciesLoading,
+  forceLoading,
+}: {
+  clientId: string;
+  dependenciesLoading: boolean;
+  forceLoading: boolean;
+} & React.HTMLAttributes<HTMLDivElement>): JSX.Element | null => {
+  const loadingOverlayState = useLoadingOverlayState(
+    clientId,
+    dependenciesLoading,
+    forceLoading
+  );
+
+  if (loadingOverlayState === 'HIDDEN') {
+    return null;
+  }
+
+  if (loadingOverlayState === 'TIMEOUT') {
+    return (
+      <div className="sp-overlay sp-error">
+        <div className="sp-error-message">
+          Unable to establish connection with the sandpack bundler. Make sure
+          you are online or try again later. If the problem persists, please
+          report it via{' '}
+          <a
+            className="sp-error-message"
+            href="mailto:hello@codesandbox.io?subject=Sandpack Timeout Error">
+            email
+          </a>{' '}
+          or submit an issue on{' '}
+          <a
+            className="sp-error-message"
+            href="https://github.com/codesandbox/sandpack/issues"
+            rel="noreferrer noopener"
+            target="_blank">
+            GitHub.
+          </a>
+        </div>
+      </div>
+    );
+  }
+
+  const stillLoading =
+    loadingOverlayState === 'LOADING' || loadingOverlayState === 'PRE_FADING';
+
+  return (
+    <div
+      className="sp-overlay sp-loading"
+      style={{
+        opacity: stillLoading ? 1 : 0,
+        transition: `opacity ${FADE_ANIMATION_DURATION}ms ease-out`,
+      }}>
+      <div className="sp-cube-wrapper" title="Open in CodeSandbox">
+        <OpenInCodeSandboxButton />
+        <div className="sp-cube">
+          <div className="sp-sides">
+            <div className="top" />
+            <div className="right" />
+            <div className="bottom" />
+            <div className="left" />
+            <div className="front" />
+            <div className="back" />
+          </div>
+        </div>
+      </div>
+    </div>
+  );
+};
+
+const useLoadingOverlayState = (
+  clientId: string,
+  dependenciesLoading: boolean,
+  forceLoading: boolean
+): LoadingOverlayState => {
+  const {sandpack, listen} = useSandpack();
+  const [state, setState] = useState<LoadingOverlayState>('LOADING');
+
+  if (state !== 'LOADING' && forceLoading) {
+    setState('LOADING');
+  }
+
+  /**
+   * Sandpack listener
+   */
+  useEffect(() => {
+    sandpack.loadingScreenRegisteredRef.current = true;
+
+    const unsubscribe = listen((message) => {
+      if (message.type === 'done') {
+        setState((prev) => {
+          return prev === 'LOADING' ? 'PRE_FADING' : 'HIDDEN';
+        });
+      }
+    }, clientId);
+
+    return (): void => {
+      unsubscribe();
+    };
+  }, [clientId, sandpack.status === 'idle']);
```

Comment:
this won’t pass the linter right? can we fix the code itself? the fix should be to make this a regular variable.

Reply:
It was just a lint warning, but I fixed it anyway. Thanks for pointing this out

Reply:
we’re moving closer to considering these equivalent in severity to compiler errors so we’d like to avoid violations on our own site. thanks for fixing!